mne_g3d/viz.py

```diff
@@ -233,3 +236,319 @@ def plot_hemisphere_mesh(vertices,
+        The size of the window, in pixels. can be one number to specify
+        a square window, or the (width, height) of a rectangular window.
+        Has no effect with mpl backend.
+    background : matplotlib color
+        Color of the background of the display window.
+    foreground : matplotlib color
+        Color of the foreground of the display window. Has no effect with mpl
+        backend.
+    initial_time : float | None
+        The time to display on the plot initially. ``None`` to display the
+        first time sample (default).
+    time_unit : 's' | 'ms'
+        Whether time is represented in seconds ("s", default) or
+        milliseconds ("ms").
+
+    Returns
+    -------
+    figure : ipyvolume.Figure
+        An instance of ipyvolume figure.
+    """
+    if not isinstance(stc, SourceEstimate):
+        raise ValueError('stc has to be a surface source estimate')
+
+    views_dict = {'lat': {'elev': 5, 'azim': 0},
+                  'med': {'elev': 5, 'azim': 180},
+                  'fos': {'elev': 5, 'azim': 90},
+                  'cau': {'elev': 5, 'azim': -90},
+                  'dor': {'elev': 90, 'azim': 0},
+                  'ven': {'elev': -90, 'azim': 0},
+                  'fro': {'elev': 5, 'azim': 110},
+                  'par': {'elev': 5, 'azim': -110}}
+
+    if views not in views_dict:
+        raise ValueError('Views must be one of ["lat", "med", "fos", "cau", '
+                         '"dor", "ven", "fro", "par"]. Got {0}.'.format(views))
+
+    subjects_dir = get_subjects_dir(subjects_dir=subjects_dir,
+                                    raise_error=True)
+    subject = _check_subject(stc.subject, subject, True)
+
+    if colorbar:
+        raise NotImplementedError('colobar=True is not yet supported.')
+
+    if not isinstance(colormap, str):
+        raise NotImplementedError('Support for "colomap" of a type other' +
+                                  ' than str is not yet implemented.')
+
+    if cortex != 'classic':
+        raise NotImplementedError('Options for parameter "cortex" ' +
+                                  'is not yet supported.')
+
+    if foreground is not None:
+        raise NotImplementedError('"foreground" is not yet supported.')
+
+    if hemi == 'split':
+        raise NotImplementedError('hemi="split" is not yet implemented.')
+
+    if hemi not in ['lh', 'rh', 'split', 'both']:
+        raise ValueError('hemi has to be either "lh", "rh", "split", '
+                         'or "both"')
+
+    scaler = 1000. if time_unit == 'ms' else 1.
+
+    if initial_time is None:
+        time_idx = 0
+    else:
+        time_idx = np.argmin(np.abs(stc.times - initial_time / scaler))
+
+    time_label, times = _handle_time(time_label, time_unit, stc.times)
+
+    stc = stc.morph(subject,
+                    grade=None,
+                    smooth=smoothing_steps,
+                    subjects_dir=subjects_dir,
+                    subject_from=subject)
```

Comment:
does this create an stc on the full resolution surface? this can create huge memory allocations. The upsampling to full resolution surface should be done on the fly as much as possible.

Reply:
as far as I understand from @larsoner comments, this operation should perform downsampling using provided parameter `smoothing_steps`.

Reply:
Reread my comment above -- I suggest to do something different, namely compute the sparse upsampling matrix (which is small in terms of memory) and do the upsampling dot operation on the fly for each time point as needed. What you have now is equivalent to doing the dot with all time points (which can be huge).

Reply:
It will work fine for a given time instance, but with enabled time viewer it will be a problem. In that case `ipyvolume` expects me to send it all data at once, for every time instance while building a mesh.  Afterwards `ipyvolume` will select the appropriate data to show.

Reply:
For many STCs this will not work. Upsampling needs to be done on the fly

Reply:
I will check with `ipyvolume` folks, but so far I suppose it will require  some changes to `ipyvolume`.

Reply:
I suppose we will need to have some callback function for time viewer

Reply:
According to [this issue](https://github.com/maartenbreddels/ipyvolume/issues/92), this problem can be solved by creating custom animation tools using ipywidgets instead of using default ones. A new mesh for each time instance should be created and the old one closed.

Reply:
It would be better just to update some data attribute rather than recreate the entire mesh